File: frontend/lib/models/notification.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'note.dart';
import 'user.dart';

part 'notification.freezed.dart';
part 'notification.g.dart';

enum NotificationType {
  mention,
  reply,
  renote,
  quote,
  reaction,
  follow,
  followRequestAccepted,
  followRequest,
  pollVote,
  pollEnded,
}

@freezed
class Notification with _$Notification {
  const factory Notification({
    required String id,
    required NotificationType type,
    required DateTime createdAt,
    User? user,
    Note? note,
    String? reaction,
    @Default(false) bool isRead,
  }) = _Notification;

  factory Notification.fromJson(Map<String, dynamic> json) =>
      _$NotificationFromJson(json);
}

File: crates/shared/src/types/notification.rs
use serde::{Deserialize, Serialize};

use super::{Note, User};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum NotificationType {
    Mention,
    Reply,
    Renote,
    Quote,
    Reaction,
    Follow,
    FollowRequest,
    FollowRequestAccepted,
    PollEnded,
    UserSignup,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Notification {
    pub id: String,
    pub created_at: String,
    pub notification_type: NotificationType,
    pub sender: Option<User>,
    pub note: Option<Note>,
    pub reaction: Option<String>,
    pub is_read: bool,
}
